### DISCUSSION TITLE
How to use a GNOME desktop entry to auto-start the Deskflow server

### DISCUSSION THREAD

**Discussion by hidara2000** [deskflow/deskflow#7591] [Show and tell]
Working perfectly on Wayland Fedora 41, Cheers.

![image](https://github.com/user-attachments/assets/cacbaf08-44c2-419d-8da8-998e5a912563)

**To enable an application menu entry and then add this to startup applications, see below.**

Create a .desktop file and add to the following path
```
~/.local/share/applications
```

Add the following:
```
[Desktop Entry]
Name=Deskflow
Version=v0.1
Icon=/home/<USER>/Applications/deskflow/res/deskflow.svg
Exec=/home/<USER>/Applications/deskflow/build/bin/deskflows -f --no-tray --debug INFO --name fedora --enable-crypto --address :24800 -c /home/<USER>/.config/Deskflow/deskflow-server.conf --tls-cert /home/<USER>/.deskflow/SSL/Deskflow.pem
Terminal=false
Type=Application
```

Note this Exec line was copied from there terminal after running Deskflow by
```
./build/bin/deskflow
```
This will be different for others and can be copied and pasted after setting up Deskflow on its first run. Also replace 	&#60;USER&#62;
I suggest running first from terminal and setting it up then run again to see the correct Exec path and flags.

Now install gnome-tweaks and add then newly available application to startup programs

![image](https://github.com/user-attachments/assets/a782d2a3-dea6-4b49-a167-3f04ec01092b)

**Comment by sithlord48** [comment]
Can we add this as a wiki page @nbolton.

**Reply by nbolton** [reply]
I'm happy to add this as a wiki page unless you want to do the honors?

**Reply by nbolton** [reply]
On second thoughts, as this is a [hack](https://github.com/deskflow/deskflow/discussions/7591#discussioncomment-10776095) (uses `synergys`) then maybe we shouldn't have it as a wiki page. 

What do you think?

**Comment by nbolton** [comment]
@hidara2000 Awesome, thanks. By the way, it's "Deskflow" with a lower case _'s'_ (not "DeskFlows"). Also heads up, I am renaming the server binary from `deskflows` to `deskflow-server` in my current WIP PR:  https://github.com/deskflow/deskflow/pull/7579 (in that PR, the packages will create a `.desktop` file)

**Reply by hidara2000** [reply]
My bad :) the binary name caught me there. Thanks for all the work, installation and setup was a breeze!

**Reply by nbolton** [reply]
No worries! When you use our Linux packages to install (coming very soon), you should see an icon like this which starts the GUI.

![image](https://github.com/user-attachments/assets/9bc8a2cd-bbfa-40fb-a12f-5d33a2786cf2)

Our `.desktop` file is at: [/res/dist/linux/deskflow.desktop](https://github.com/deskflow/deskflow/blob/master/res/dist/linux/deskflow.desktop)

However, it will be moved to a new place in PR #7579

**Reply by nbolton** [reply]
One thought... I wonder if perhaps your new `.desktop` entry you have created should have the name "Deskflow Server" to indicate that it's the server binary starting instead of the GUI?

**Reply by nbolton** [reply]
This gives me an idea... maybe our package could create `.desktop` files for the server and client which uses a config file for the args (we have a new `.toml` config reader where you can put the args) rather than having the server args hard coded in the `.desktop` file.

**Reply by sithlord48** [reply]
.desktop files are not for non gui apps.

what should be done is a proper service using systemd

**Reply by nbolton** [reply]
That's very true. I suppose in that case the original suggestion is a hack:
```
Exec=/home/<USER>/Applications/deskflow/build/bin/deskflows -f --no-tray --debug INFO --name fedora --enable-crypto --address :24800 -c /home/<USER>/.config/Deskflow/deskflow-server.conf --tls-cert /home/<USER>/.deskflow/SSL/Deskflow.pem
```

**Reply by sithlord48** [reply]
it is, but its also a common one that end users will do on a system where you can't select items for autostart without  that item having a menu entry.

**Comment by hidara2000** [comment]
I'm having a play to see if I can get a python script to create the services and .desktop (see attached)

SERVER
```
python3 ./init_deskflow.py "/home/$USER/Applications/deskflow/build/bin/deskflows -f --no-tray --debug INFO --name $(hostname -f) --enable-crypto --address :24800 -c /home/$USER/.config/Deskflow/deskflow-server.conf --tls-cert /home/$USER/.deskflow/SSL/Deskflow.pem"
```

CLIENT
```
python3 ./init_deskflow.py "/home/$USER/Applications/deskflow/build/bin/deskflowc -f --no-tray --debug INFO --name $(hostname -f) --enable-crypto --sync-language 192.168.0.244:24800"  
```

The .desktop uses the deskflow binary now instead of the deskflows

I am getting the following errors
![image](https://github.com/user-attachments/assets/8e7b28ce-a4f6-453b-8bd7-a95efe7afed2)

.service file
![2024-09-28_13-22](https://github.com/user-attachments/assets/db59b82d-3954-448b-bae5-25baeb8165ba)



[init_deskflow.zip](https://github.com/user-attachments/files/17172080/init_deskflow.zip)

**Reply by nbolton** [reply]
Interesting. Thanks for sharing!

**Reply by sithlord48** [reply]
You don't seam to be using the EI backend . Check the --help you may need to add an option like `use-ei`

**Reply by sithlord48** [reply]
You also may need to run this as a user service to be able to provide portal access

**Reply by nbolton** [reply]
In Deskflow EI is automatically used if Wayland is detected (works a bit differently to Input Leap which is manual and explicit, I believe). I guess we need to add an arg for environments other than Wayland. Maybe Deskflow should automatically use EI if X is not detected (rather than if Wayland is detected)...
